File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        Map = {")": "(", "]": "[", "}": "{"}
        stack = []

        for bracket in s:
            if bracket not in Map:
                stack.append(bracket)
                continue
            if not stack or stack[-1] != Map[bracket]:
                return False
            stack.pop()

        return not stack
        

File: 0020-valid-parentheses/0020-valid-parentheses.cs
public class Solution {
    public bool IsValid(string s) {
        Dictionary<char, char> brackets = new Dictionary<char, char>{ 
            {')', '('}, 
            {'}', '{'}, 
            {']', '['} 
        };
        Stack<char> stack = new Stack<char>();
        
        foreach(char bracket in s) {
            if (brackets.ContainsKey(bracket)) {
                if (stack.Count > 0 && brackets[bracket] == stack.Peek()) {
                    stack.Pop();
                } else {
                    return false;
                }
            } else {
                stack.Push(bracket);
            }
        }
        return stack.Count == 0;
    }
}